Tx evaluation

ginnun opened the pull request:
Resolves #306

**Stacked on #537** - merge that one first.

## Context

Implements tx evaluation APIs compatible with the Blockfrost.io version:

- `/utils/txs/evaluate`
- `/utils/txs/evaluate/utxos`

## What is added

- Two new endpoints with integration tests
- `bf-tx-evaluator` crate providing native (Rust, via `pallas-validate`) and external (Haskell, via `testgen-hs`) evaluators
- `ChainConfigWatch` in `bf-node` - a background task that watches the Cardano node and publishes protocol params + genesis config via `tokio::sync::watch`, refreshed only at epoch boundaries
- Node helpers: `ledger`, `utxo`, `chain_config_watch`

## What is changed

- Pallas fork switched to `blockfrost/pallas` with `pallas-validate` (`phase2` feature) for script evaluation
- CBOR validation error messages improved (graceful fallback on `pallas-hardano` formatting failure)
- Devshell picks up tools needed by the new native deps - m4 (for `secp256k1-sys`), make/awk/diffutils (for integration tests)
- `cardano-node` trace silencing migrated to the newer API
- CI: `CARGO_PROFILE_DEV_DEBUG=0` on integration-test jobs to reduce disk pressure

## What is scoped out (now in #537)

- `testgen-hs` subprocess handling extracted from `crates/node/src/cbor/` into dedicated `bf-testgen` and `bf-error-decoder` crates
- `testgen-hs` version bump to `10.6.3.1` and release binary downloads switched to `blockfrost/testgen-hs`
- `build_utils` target detection helpers factored out

michalrus commented on nix/internal/unix.nix:
```diff
@@ -163,14 +166,15 @@ in
     in
       (import inputs.flake-compat {
         src =
+          if targetSystem != "aarch64-darwin" && targetSystem != "aarch64-linux"
```
This looks like a merge/rebase artifact – wrong conflict resolution? The line is repeated.

michalrus commented on nix/internal/unix.nix:
```diff
@@ -220,7 +224,24 @@ in
+              cp -r ${unpatched} $out
+              chmod -R +w $out
+              cd $out
+              patch -p1 -i ${./testgen-hs--enable-aarch64-linux.diff}
```
Wait, but we already have `aarch64-linux` on `main`. Something is off here, another conflict?

michalrus commented on src/api/utils/txs/evaluate/model.rs:
```diff
@@ -0,0 +1,359 @@
+        let _req: TxEvaluationRequest = serde_json::from_value(value).unwrap();
+    }
+}
+*/
```
Should this file be commented out and committed?

It also looks like we don't have `treefmt` run on this, since it's missing a trailing newline.

ginnun replied:
Fixed in b4b535e552d52a2541d81728db47811dc9b1ace8

michalrus commented on crates/testgen/src/testgen.rs:
```diff
@@ -11,19 +11,22 @@ use tokio::sync::{mpsc, oneshot};
-    sender: mpsc::Sender<FDRequest>,
-    current_child_pid: Arc<AtomicU32>,
+pub struct Testgen {
+    pub sender: mpsc::Sender<TestgenRequest>,
```
Why do these internal fields need to be public for everyone?

ginnun replied:
Changed in bbec0dbf45875089fe67ec1e08faf79ab47aacac

michalrus commented on crates/testgen/src/testgen.rs:
```diff
@@ -143,36 +151,6 @@ impl FallbackDecoder {
     }
 
-    /// This function is called at startup, so that we make sure that the worker is reasonable.
-    pub async fn startup_sanity_test(&self) -> Result<(), String> {
```
But isn’t this useful? Just in case that someone plugs another binary into the `libexec` directory.

ginnun replied:
This uses error serialization/deserialization. I can expect every binary integration to have a startup_sanity_test implementation of their own.
For evaluation, it's already handled via the init functionilty.

michalrus replied:
I was thinking more like… What if the user puts a different binary there? It was more of a sanity test from the point of view of the `blockfrost-platform`, i.e. "is this a reasonable helper binary that we expect" to rule out obvious user errors

michalrus reviewed (commented):
Hey, I started reviewing, but then noticed that it's a draft! Left a few comments either way :)

michalrus reviewed (commented):
I pushed another merge with `main`, because there were some more conflicts to resolve. I hope I did that correctly!

Now I see that `cargo test` tests are failing:

```
❯ cargo nextest run --workspace --lib --verbose

…

     Summary [  76.650s] 650 tests run: 195 passed, 455 failed, 0 skipped
```

Stuff like:

```
        FAIL [   1.058s] blockfrost-platform-error-decoder external::tests::test_cbor_0001_ConwayTreasuryValueMismatch_Coin_Coin
```

Or:

```
        FAIL [   1.086s] blockfrost-platform-error-decoder tests::specific::test_cbor_0013_ConwayWdrlNotDelegatedToDRep_KeyHash_KeyHash
```

michalrus commented:
> The plan is to fix/update the code for `pallas/hardano`.

Great! Here's a tracking issue for this new work:
* #459

michalrus commented on nix/internal/unix.nix:
```diff
@@ -191,9 +191,28 @@ in
     cardano-node-packages =
       {
-        x86_64-linux = cardano-node-flake.hydraJobs.x86_64-linux.musl;
+        x86_64-linux = cardano-node-release-linux;
```
And since the tests are passing – this would mean that SREs updated their node on our GitHub Actions runner.

michalrus replied:
I think I get it now – please see https://github.com/blockfrost/blockfrost-platform/pull/378#discussion_r2872404626

michalrus commented on nix/internal/unix.nix:
```diff
@@ -188,9 +191,28 @@ in
     cardano-node-packages =
       {
-        x86_64-linux = cardano-node-flake.hydraJobs.x86_64-linux.musl;
+        x86_64-linux = cardano-node-release-linux;
```
I'm fairly certain that this is not necessary in this PR, let's verify again.

See these comments: https://github.com/blockfrost/blockfrost-platform/pull/378#discussion_r2872247931.

michalrus reviewed (commented):
WDYT of this last change before a proper review, so that `testgen-hs` version is in sync with `cardano-node` in the shell? Hopefully it's only a formality, but worth checking, IMO.

hlolli commented on .github/workflows/ci.yaml:
```diff
@@ -127,6 +127,9 @@ jobs:
         uses: actions/checkout@v4
 
+      - name: Free disk space
+        run: nix-collect-garbage -d
```
It's correct that the runner doesn't always clean the repos after pulling them. Deleting build folder could help a little if it's big. I think the checkout action pulls just 1 commit and doesn't store history, but doesn't remove the repo afterwards, but removes and cleans on fetch

<img width="1514" height="248" alt="image" src="https://github.com/user-attachments/assets/320de71f-4f08-41d8-8e58-79dd85de9c02" />

ginnun replied:
> Other PRs don’t fill it, so maybe it's testgen-hs somehow?
@michalrus   It's the Rust debug build artifacts accumulating in the runner's /run/github-runner/.../target/debug/deps/ directory (after my /nix/store clean), not testgen-hs.
                                                                                                                                                                                                                                              
  This PR adds pallas_validate as a new transitive dependency (via tx_evaluator). In debug mode (default debug = 2), its .rlib is hundreds of MB. From what I understand (i might be wrong), the self-hosted runner never cleans target/, so old hashed artifacts from previous runs pile up.
  The linker failure tells that it dies (just after tx_evaluator/pallas_validate) trying to write the final test binary after the Nix GC already freed 16 GB:
  ld.bfd: final link failed: No space left on device

  Temporary fix to check the theory: clean target/debug/ before building + set CARGO_PROFILE_DEV_DEBUG=0 (maybe later) to keep future artifacts small. Working on it.

michalrus replied:
@hlolli, @cleverca22, I mean, but should we then be running the builds inside `/run/` (on `tmpfs`, in limited RAM), instead of somewhere on the (more unlimited) SSD?

Maybe the GitHub runner service is configured in a strict systemd sandbox which defaults to `/run`? Or are its directories in `/run` just symlinks to somewhere on the SSD?

michalrus commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,893 @@
+
+pub fn create_address(addr: &str) -> Bytes {
+    Address::from_bech32(addr)
+        .unwrap_or_else(|_| Address::from_hex(addr).unwrap())
```
User-supplied strings should not cause `panic!`s

michalrus commented on crates/platform/src/api/utils/txs/evaluate/utxos.rs:
```diff
@@ -1,6 +1,102 @@
+                        evaluate_binary_tx(node, tx_cbor.as_slice(), request.additional_utxo_set)
+                            .await?,
+                    );
+                    serde_json::to_value(r).unwrap()
```
Can this `panic!`? Let's use `?`, too

michalrus commented on crates/platform/src/server/routes/hidden.rs:
```diff
@@ -135,9 +136,9 @@ pub fn get_hidden_api_routes(enable_metrics: bool) -> Router<AppState> {
-        .route("/utils/tx/evaluate", post(utils::txs::evaluate::root::route))
-        .route("/utils/tx/evaluate/utxos", post(utils::txs::evaluate::utxos::route));
+        // tx evaluate
+        .route("/utils/tx/evaluate",post(utils::txs::evaluate::root::route))
```
The original legacy Blockfrost API uses plural here (`/txs/`, not `/tx/`):

https://docs.blockfrost.io/#tag/cardano--utilities/POST/utils/txs/evaluate

<img src="https://github.com/user-attachments/assets/74257f4f-b965-43d9-adaa-9f465692a23c" />

michalrus commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1000 @@
+
+// JSON response
+#[derive(Serialize)]
+pub struct TxEvaluationResponse {
```
Is it enough that it only contains `spend:1` and `mint:0`?

Also: dead code?

michalrus commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,893 @@
+    tx_cbor_binary: &[u8],
+    utxos: Option<AdditionalUtxoSet>,
+) -> Result<EvalReport, BlockfrostError> {
+    let slot_config = pallas_validate::phase2::script_context::SlotConfig::default();
```
Could it be that the default `SlotConfig` wouldn’t match the network’s?

ginnun replied:
Derived SlotConfig from genesis in e1be8f1.

michalrus reviewed (changes requested):
Ahh, it’s a large one! 🐳

Some comments below.

Can we also add the two new endpoints to the ones we run `blockfrost-tests` against in `supported_endpoints.json`? It will be automatically picked up by the CI, and we can see the change:

https://github.com/blockfrost/blockfrost-platform/blob/db42ec026433a6958e608c2a4bd2130880417b3a/crates/platform/tests/data/supported_endpoints.json#L1-L4

ginnun commented on crates/common/src/chain_config.rs:
```diff
@@ -0,0 +1,76 @@
+    pub fn by_network_magic(network_magic: &u32) -> Self {
+        match network_magic {
+            764824073 => Self::preview(),
+            1 => Self::mainnet(),
```
Yeah, how did I do this? These are only used in slot calculations, maybe this is why they slipped. 

I couldn't find a common place to use them from.

Fixed in bdad9bd356669dc66c98fc679906fcd8f5991f39 .

ginnun commented on crates/common/src/errors.rs:
```diff
@@ -70,6 +70,12 @@ impl From<hex::FromHexError> for BlockfrostError {
 
+impl From<serde_json::Error> for BlockfrostError {
+    fn from(err: serde_json::Error) -> Self {
+        Self::internal_server_error(format!("SerdeError: {err}"))
```
Fixed in 3e17ec8.

ginnun commented on crates/common/src/chain_config.rs:
```diff
@@ -0,0 +1,76 @@
+            genesis_config,
+            protocol_params,
+            slot_config,
+            era: 6, // conway
```
Added a comment in 7cebed7.

michalrus replied:
Out of curiosity, what happens if:
- the node is still syncing, and say, still in Babbage?
- they add era № 7, will we fail silently?

ginnun commented on crates/tx_evaluator/src/ledger.rs:
```diff
@@ -0,0 +1,11 @@
+pub enum TransactionScriptFailure {
```
Removed in 3e1748f.

ginnun commented on .gitignore:
```diff
@@ -31,6 +31,7 @@ result-*
 
 /cardano-node-configs
 /cardano-node
+/data
```
I used this folder for cardano and dolos node data (the way I prefer), also as an output folder for various bash commands.
Once creates a problem, this kind of things can always be changed on spot.
Removed it in ab57e66 anyways.

michalrus replied:
For a personal Git ignore, you also have:
- `~/.config/git/ignore` – global,
- `.git/info/exclude` – in this repo

ginnun commented on crates/testgen/src/testgen.rs:
```diff
@@ -85,52 +124,92 @@ impl FallbackDecoder {
+        }
+
+        // Compile-time CARGO_MANIFEST_DIR always points to this crate.
+        let target_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(target_dir_from_env);
```
Applied in 56996fc.

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,893 @@
+            .iter()
+            .map(convert_alonzo_txin)
+            .collect(),
+        _ => unreachable!("unknown era transaction"),
```
Converted into a 400 error in e43824c4b898f709bf3a1949102133a21cf718d1.

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,915 @@
+
+fn parse_asset_string(str: &str) -> (PolicyId, AssetName) {
+    let mut parts = str.split('.');
+    let policy_id = parts.next().expect("Invalid asset string in utxo output");
```
@hlolli Are you suggesting to keep the line as is?

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,893 @@
+    }
+    let mut node = node_pool.get().await?;
+
+    let txins = extract_inputs(multi_era_tx.clone());
```
Fixed in cb3cad6.

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,893 @@
+            script_ref,
+        };
+
+        let tx_out_cbor = pallas_codec::minicbor::to_vec(post_alonzo).unwrap();
```
These `unwraps` won't fail since they are round-trips and already valid pallas data.
To create a `KeepRaw<>` object, you need to provide raw bytes, so we are round-tripping.
But to get rid of future suspicions, applied your feedback in 8738486ff3bb5ac26b963b6ac94dd46bb3f3d9f1.

ginnun commented on crates/platform/src/api/utils/txs/evaluate/utxos.rs:
```diff
@@ -1,6 +1,102 @@
+    // Allow only application/json content type
+    validate_content_type(&headers, &["application/json"])?;
+
+    let version: u8 = query.version.parse().unwrap();
```
Handled in 8738486ff3bb5ac26b963b6ac94dd46bb3f3d9f1.

ginnun commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1000 @@
+            ScriptV6::PlutusV3 { cbor } => ScriptRef::PlutusV3Script(PlutusScript::<3>(
+                Bytes::from(hex::decode(cbor).unwrap()),
+            )),
+            ScriptV6::Native { cbor, json } => cbor.clone().map_or_else(
```
cbor and json coming from pallas is not possible. Although there are both `Option`, it's in xor fashion.

But should you trust an external lib?

Handled the worst case in d6f0b64455974f75d1a4e5b799b739cbadad22ac.

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,145 @@
+        serde_json::to_value(r)
+            .unwrap()
+            .as_object()
+            .unwrap()
```
I am not sure how can these already type-safe & serde decorated data can fail? 
But still, I converted them to more descriptive `.expect` 00fc720

ginnun commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,262 @@
+            AppError::Server(format!("ExternalEvaluator: Failed to initialize: {err}"))
+        })
+        */
+        // println!("1 -> {}", payload);
```
Removed in 00fc720.

ginnun commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,262 @@
+        Ok(evaluator)
+    }
+
+    /// Sends repeatative data as the first communication so we don't need to send every time.
```
Corrected in 00fc720.

ginnun commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1000 @@
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum TxValidatorV6 {
```
Yes, we will get only these two in an evaluation response.
I removed this.

ginnun commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1000 @@
+    Deserialization(DeserializationErrorData),
+}
+
+pub enum EvaluateTransactionError {}
```
Removed in 44e7945.

ginnun commented on crates/error_decoder/src/external.rs:
```diff
@@ -0,0 +1,1537 @@
+    }
+
+    #[cfg(test)]
+    /// A single global [`ExternalDecoder`] that you can cheaply use in tests.
```
Handled in bdad9bd.

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,145 @@
+    match resp {
+        TestgenResponse::Ok(value) => {
+            println!("value6 {:?}", value);
+            let decoded: Vec<TxEvalResultV6> = serde_json::from_value(value).unwrap();
```
TestgenResponse::ok already contains a `serde::value` here, so there should be no way for it to fail serialization.
I used more descriptive `.expect` in https://github.com/blockfrost/blockfrost-platform/commit/00fc720ba3ac5e381c2732b339f31cf35b875332

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,145 @@
+                    json!({
+                        "jsonrpc": "2.0",
+                        "method": "evaluateTransaction",
+                        "error": "hello error",
```
Fixed in [c901a81](https://github.com/blockfrost/blockfrost-platform/pull/378/commits/c901a81).

ginnun commented on crates/build_utils/src/target.rs:
```diff
@@ -0,0 +1,21 @@
+pub fn os() -> &'static str {
```
Looks like a rebase/merge artifact. Deleted in [f269525](https://github.com/blockfrost/blockfrost-platform/pull/378/commits/f269525).

ginnun commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1000 @@
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum TxEvalResultV5 {
+    SUCCESS(TxEvalSuccessV5),
```
Corrected in 00fc720ba3ac5e381c2732b339f31cf35b875332.

ginnun commented on crates/platform/tests/common/mod.rs:
```diff
@@ -30,6 +31,14 @@ pub fn initialize_logging() {
     let _ = INIT_LOGGING;
 }
 
+#[allow(dead_code)]
```
Removed in https://github.com/blockfrost/blockfrost-platform/commit/e15aa7dc6dc2be4ba4c89d5016d3d2fbd7b66415.

hlolli commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,915 @@
+            let mut assets = BTreeMap::new();
+            for (id_name, number) in assets_map {
+                let mut asset_detail = BTreeMap::new();
+                let coin: PositiveCoin = number
```
No, endpoint is reachable via query override; malformed assets/policy IDs can panic.
I could suggest we replace expect/unwrap in this path with fallible validation (map_err(...)?) and return BlockfrostError::custom_400(...) for malformed additionalUtxo fields (asset id/name/amount).
If this is experimental, you can ignore this comment :)

hlolli commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,155 @@
+        TestgenResponse::Ok(value) => {
+            // The external evaluator returns a v6-like structure, but we need to convert it to v5.
+            let v6_results: Vec<TxEvalResultV6> =
+                serde_json::from_value(value).expect("ExternalEvaluator: invalid v6 result format");
```
no a big deal, more descriptive .expect is also good for debugging and reliability

hlolli commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1005 @@
+            ScriptNative::ExpiresAt(time) => NativeScript::InvalidHereafter(time),
+            ScriptNative::StartsAt(time) => NativeScript::InvalidBefore(time),
+            ScriptNative::NOf(h_map) => {
+                let (n_str, scripts) = h_map.into_iter().next().expect("NOf: empty map");
```
To elaborate: my concern is not about valid requests, but about what happens on malformed request bodies in the runtime request path.

In this code, `ScriptNative::NOf` is modeled as `HashMap<String, Vec<ScriptNative>>`, and then later converted with:

 ```rust
 let (n_str, scripts) = h_map.into_iter().next().expect("NOf: empty map");
 n_str.parse::<u32>().expect("NOf: invalid n key")
 ```

 That means some bad payloads can make us panic instead of returning a deterministic `400`.

 Concrete examples:

 ```json
 { "native": {} }
 ```
 can hit `expect("NOf: empty map")`

 ```json
 { "native": { "foo": [] } }
 ```
 can hit `expect("NOf: invalid n key")`

 and similarly nearby helpers like `decode_pubkey_hash` / `decode_script_hex` use `expect(...)`, so malformed key hashes or script CBOR can also panic.

 Since this is request parsing/normalization for client input, I’d prefer either:

 1. reject these forms during deserialization, or
 2. make the conversion fallible (`TryFrom` / `Result<..., BlockfrostError>`) and map invalid input to `custom_400(...)`.

The main point is: invalid client payload should become a controlled `400`, not a process panic / `500`.

hlolli commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,267 @@
+                    Ok(Self::build_transaction_output(
+                        txin,
+                        create_address(&tout.address)?,
+                        tout.script
```
perfect

michalrus commented on crates/node/src/cbor/tests/test_data.rs:
```diff
@@ -0,0 +1,185 @@
+use pallas_primitives::{
```
This file is never included in compilation?

michalrus commented on crates/platform/src/api/utils/txs/evaluate/utxos.rs:
```diff
@@ -1,6 +1,75 @@
+    extract::Json(tx_request): extract::Json<TxEvaluationRequest>,
+) -> Result<impl IntoResponse, BlockfrostError> {
+    // Allow only application/json content type
+    validate_content_type(&headers, &["application/json"])?;
```
Is this necessary? We already define the parameter as `extract::Json(tx_request): extract::Json<TxEvaluationRequest>`, so we won’t even get to this line unless it was indeed JSON.

michalrus reviewed (changes requested):
A couple more comments:

ginnun commented on crates/testgen/src/testgen.rs:
```diff
@@ -0,0 +1,325 @@
+                );
+                let restart_delay = std::time::Duration::from_secs(1);
+                error!(
+                    "Testgen: will restart in {restart_delay:?} because of a subprocess error: {}",
```
When it comes to tx evaluation, It's not a fallback anymore. It's an external evaluator, The name "external" is better in my opinion. Whatever we use it as a fallback or a main is up to us and should not be enforced in the name Hence, it's now the binaries name.

michalrus reviewed (changes requested):
One quick comment about the other repo (Pallas), and now I'll process the rest (here):

michalrus commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1041 @@
+    }
+}
+
+impl<'de> Deserialize<'de> for DeserializationErrorData {
```
Hmmmm, you're _deriving_ `Serialize` above:

https://github.com/blockfrost/blockfrost-platform/blob/fe4dd66ee371471e784202e9b94cfb1c15627000/crates/tx_evaluator/src/model/api.rs#L184-L185

… so after this generic serialization, it ends up as JSON object with 6 fields.

But here, it seems like you're deserializing a single string?

So serializer is not a inversion of deserializer, and they don’t match? I think in general JSON codecs should be invertible, or it’s surprising.

michalrus reviewed (changes requested):
Okay – another round done! The most important comment is about `blockfrost-tests`, I think.

michalrus commented on crates/platform/tests/data/supported_endpoints.json:
```diff
@@ -37,5 +37,7 @@
   "/addresses/{address}/transactions",
-  "/addresses/{address}/utxos"
+  "/addresses/{address}/utxos",
+  "/utils/txs/evaluate",
```
And you can easily run a decent Dolos with `run-dolos-preview` inside the devshell.

But since you don’t need to test any Dolos integration, you can just as well comment it out!

michalrus replied:
> I should underline, we don't have mempool integration yet (Ogmois does use mempool + chain)

- So the implementation is not equivalent, even though we're deferring to Haskell code? 
- Shouldn’t we add it before releasing? 👀 
- How hard is it?
- And why isn’t it caught by [`blockfrost-tests`](https://github.com/blockfrost/blockfrost-tests)?

ginnun replied:
> So the implementation is not equivalent, even though we're deferring to Haskell code?

Yes, but it's not related to Rust or Haskell. It's about what we feed to evaluation logic.
> Shouldn’t we add it before releasing? How hard is it?

When this implementation was completed, pallas didn't have capability to fetch mempool from the node. This is why it is missing. I don't think this should be a blocker.
> And why isn’t it caught by blockfrost-tests?

We don't have a test case that can catch this.

ginnun commented on Cargo.toml:
```diff
@@ -52,13 +57,16 @@ metrics-exporter-prometheus = { version = "0.18.1", default-features = false }
+pallas-addresses = { git = "https://github.com/blockfrost/pallas.git", tag = "blockfrost-platform-0.0.3-alpha6" }
+pallas-codec = { git = "https://github.com/blockfrost/pallas.git", tag = "blockfrost-platform-0.0.3-alpha6" }
+pallas-crypto = { git = "https://github.com/blockfrost/pallas.git", tag = "blockfrost-platform-0.0.3-alpha6" }
+pallas-hardano = { git = "https://github.com/blockfrost/pallas.git", tag = "blockfrost-platform-0.0.3-alpha6" }
```
@michalrus https://github.com/txpipe/pallas/pull/624 is ready.

ginnun reviewed (commented):
@michalrus I replied, but kept them in `pending` for a while ^^

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,165 @@
+        "version": "1.0",
+        "servicename": "ogmios",
+        "methodname": "EvaluateTx",
+        "result": { "EvaluationResult": result_map },
+        "reflection": reflection,
+    })
+}
+
+pub fn wrap_response_v6(resp: TestgenResponse) -> serde_json::Value {
+    let id = serde_json::json!(uuid::Uuid::new_v4().to_string());
+    match resp {
+        TestgenResponse::Ok(value) => {
+            let decoded: Vec<TxEvalResultV6> =
+                serde_json::from_value(value).expect("ExternalEvaluator: invalid v6 result format");
+
```
Fixed in 359d8b12.

ginnun commented on crates/platform/src/server.rs:
```diff
@@ -61,6 +62,12 @@ pub async fn build(
+    let chain_config_cache = init_caches(node_conn_pool.clone()).await?;
+
+    // Initialize the Haskell-based tx evaluator
+    let tx_evaluator = ExternalEvaluator::spawn(chain_config_cache).await?;
```
This is known and handled in another issue/pr: https://github.com/blockfrost/blockfrost-platform/pull/490

ginnun commented on crates/testgen/src/testgen.rs:
```diff
@@ -0,0 +1,310 @@
+    pub fn spawn(variant: &str) -> Result<Self, AppError> {
+        let testgen_hs_path = Self::find_testgen_hs().map_err(AppError::Server)?;
+
+        info!(
+            "Using {} as a fallback CBOR error decoder",
+            &testgen_hs_path
+        );
```
Not a bad idea, no harm to log more info here. Fixed in 359d8b12.

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,165 @@
+    json!({"id": uuid::Uuid::new_v4().to_string()})
+}
+
+/// V5 responses does generate three kind of json:
+/// - successfull eval; top-level 'result' field with exec units
+/// - failed eval: top-level 'result' field with error json details
+/// - faulty input: top-level 'fault' field with error string
+pub fn wrap_response_v5(resp: TestgenResponse) -> serde_json::Value {
+    let reflection = generate_reflection();
+    match resp {
+        TestgenResponse::Ok(value) => {
+            // The external evaluator returns a v6-like structure, but we need to convert it to v5.
+            let v6_results: Vec<TxEvalResultV6> =
+                serde_json::from_value(value).expect("ExternalEvaluator: invalid v6 result format");
+
```
Fixed in 359d8b12.

ginnun commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,165 @@
+    json!({"id": uuid::Uuid::new_v4().to_string()})
+}
+
+/// V5 responses does generate three kind of json:
+/// - successfull eval; top-level 'result' field with exec units
+/// - failed eval: top-level 'result' field with error json details
+/// - faulty input: top-level 'fault' field with error string
```
Fixed in 359d8b12.

ginnun commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1121 @@
+    }
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Value {
+    pub coins: u64,
+    pub assets: Option<HashMap<String, u64>>, // asset name and number. Asset number can be negative when burning assets but this behaviour changed in Conway. Now it can be only PositiveCoin
+}
```
Fixed in 359d8b12.

ginnun commented on crates/platform/src/api/utils/txs/evaluate/utxos.rs:
```diff
@@ -1,6 +1,71 @@
+    hex::decode(hex_str).map_err(|e| {
+        BlockfrostError::custom_400(format!("invalid hex-encoded transaction CBOR: {e}"))
+    })
 }
```
Correct, I added base64 support back to these APIs in 8a7184fbe36f3c2282d54f2dbfea23358dcd65a1

ginnun commented on crates/common/src/config.rs:
```diff
@@ -28,6 +28,7 @@ pub struct Config {
     pub network: Network,
     pub custom_genesis_config: Option<PathBuf>,
     pub data_node: Option<DataNodeConfig>,
+    pub evaluator: Evaluator,
```
Removed in 29c32398.

ginnun commented on crates/common/src/helpers.rs:
```diff
@@ -13,7 +14,10 @@ pub fn binary_or_hex_heuristic(xs: &[u8]) -> Vec<u8> {
 
     if !even_length || contains_non_hex {
-        xs.to_vec()
+        use base64::{Engine as _, engine::general_purpose};
```
No need to check this, I reverted my changes in this function and implemented a new one.

michalrus commented:
BTW, I found this in my notes:

> Re: https://github.com/blockfrost/blockfrost-platform/pull/378#discussion_r2923147312
>
> tx-eval shouldn’t need to know about HTTP errors

Is it still applicable?

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,977 @@
+    let conway_pp: ConwayProtParams = ConwayProtParams {
+        minfee_a: pp.minfee_a.unwrap() as u32,
+        minfee_b: pp.minfee_b.unwrap() as u32,
+        max_block_body_size: pp.max_block_body_size.unwrap() as u32,
```
I removed `unwrap`s in f1ad42109fac16397d1dab705ad71afc661ec702

ginnun commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,977 @@
+        convert_bigint(&sys_start.year)? as i32,
+        sys_start.day_of_year as u32,
+    )
+    .expect("Invalid system start date");
```
I got rid of `unwrap`s in f1ad42109fac16397d1dab705ad71afc661ec702.

ginnun commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,268 @@
+            ))
+        })?;
+
+        let response = self.testgen.send(json).await.map_err(|err| {
```
Sure, added in 7d926630.

ginnun commented:
> BTW, I found this in my notes:
> 
> > Re: [#378 (comment)](https://github.com/blockfrost/blockfrost-platform/pull/378#discussion_r2923147312)
> > tx-eval shouldn’t need to know about HTTP errors
> 
> Is it still applicable?

Nope, related part (unwrap) to [#378 (comment)](https://github.com/blockfrost/blockfrost-platform/pull/378#discussion_r2923147312) is completely removed.

But I don't remember the context about HTTP errors.

ginnun commented:
I can make the CI green after https://github.com/blockfrost/testgen-hs/pull/9 and https://github.com/blockfrost/blockfrost-tests/pull/65 are merged and a new `testgen-hs` is released.

michalrus reviewed (changes requested):
Note to self for when this PR is close to a merge:

> Let's have a small PR to move `testgen-hs` handling on `main` in accordance to what the `tx-eval` PR did – for better view of conflicts and what _really_ changed

(requesting changes so that it's not lost)

ginnun commented on crates/tx_evaluator/src/ogmios5_response.rs:
```diff
@@ -0,0 +1,503 @@
+            .unwrap_or(message),
+    );
+
+    json!({
+        "EvaluationFailure": {
+            name.to_owned(): {
+                "reason": reason
+            }
```
This compiles and produces the intended key. I verified it via a standalone test.

ginnun commented on crates/node/src/chain_config_watch.rs:
```diff
@@ -0,0 +1,212 @@
+    let current_slot = slot_config.zero_slot + elapsed_ms / slot_config.slot_length;
+
+    // Slots into the current epoch.
+    let slot_in_epoch = (current_slot - slot_config.zero_slot) % slot_config.epoch_length;
+    let slots_until_next = (slot_config.epoch_length - slot_in_epoch) % slot_config.epoch_length;
+
+    let ms_until_next = slots_until_next * slot_config.slot_length;
+    Duration::from_millis(ms_until_next) + EPOCH_BOUNDARY_BUFFER
```
Fixed in 111860b3.

ginnun commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,408 @@
+        }
+
+        if !overlapping.is_empty() {
+            return Err(BlockfrostError::custom_400(format!(
+                "Additional UTxO set contains UTxOs that conflict with on-chain state: {}",
+                overlapping.join(", ")
+            )));
```
Handled in 111860b3.

ginnun commented:
> Note to self for when this PR is close to a merge:
> 
> > Let's have a small PR to move `testgen-hs` handling on `main` in accordance to what the `tx-eval` PR did – for better view of conflicts and what _really_ changed
> 
> (requesting changes so that it's not lost)

@michalrus This is in the making.

ginnun committed:
feat(tx-evaluation): squashed work for stacked rebase

- chore: add devshell tools and CI tweaks for tx evaluation
- chore: update workspace deps for tx evaluation
- feat(common): add lazy chain config loader and helpers
- fix(node): improve CBOR validation error messages
- feat(node): add chain config, ledger, and utxo helpers
- feat: add bf-tx-evaluator crate with native and external evaluators
- feat(platform): add /utils/txs/evaluate endpoints
- test: add coverage for tx evaluate endpoints

ginnun commented:
Failing tests are unrelated to this PR.

For evaluate failure, please see https://github.com/blockfrost/blockfrost-tests/pull/73

ginnun committed:
chore: bump blockfrost-tests flake input

ginnun committed:
Merge remote-tracking branch 'origin/main' into feat/tx-evaluation-rebased

michalrus commented on Cargo.toml:
```diff
@@ -67,13 +69,16 @@ nix = { version = "0.30", default-features = false, features = ["signal"] }
-pallas-network = { git = "https://github.com/txpipe/pallas.git", rev = "ed422bc0d5a09f77617d4882dd4c71089a8bcdb2" }
-pallas-primitives = { git = "https://github.com/txpipe/pallas.git", rev = "ed422bc0d5a09f77617d4882dd4c71089a8bcdb2" }
-pallas-traverse = { git = "https://github.com/txpipe/pallas.git", rev = "ed422bc0d5a09f77617d4882dd4c71089a8bcdb2" }
+pallas-addresses = { git = "https://github.com/blockfrost/pallas.git", rev = "cfa0c633640f5989e44f50ed27efc38992f9ed0d" }
```
There has been one more commit since then in your PR – https://github.com/txpipe/pallas/pull/624/changes/9a83c74125738d198010dcb4d61a3595a6d4d771

ginnun replied:
Updated in c26479f542ae74bb00e675425fa6482a13673df6

michalrus commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,977 @@
+
+        utxos.insert(TxoRef::from(&multi_era_in), EraCbor::from(multi_era_out));
+    }
+    let mut node = node_pool.get().await?;
```
I don't see this comment here. Maybe more like `FIXME`, not `TODO`? 🫣

ginnun replied:
Ah sorry for the confusion. I think I fixed this while I was implementing the `ChainConfigWatch`. Line 79 reads cached values and later they are fed into here.

```Rust
    let config = config_watch.get()?;
```

michalrus commented on crates/node/src/chain_config_watch.rs:
```diff
@@ -0,0 +1,233 @@
+        let new_config = loop {
+            match init_caches(node_pool.clone()).await {
+                Ok(c) => break c,
+                Err(e) => {
+                    tracing::warn!(
+                        "ChainConfigWatch: failed to refresh chain config: {e}. Retrying in {}s.",
+                        RETRY_INTERVAL.as_secs()
+                    );
+                    tokio::select! {
+                        () = time::sleep(RETRY_INTERVAL) => {},
+                        () = tx.closed() => {
+                            tracing::info!("ChainConfigWatch: stopping monitor");
+                            return;
+                        },
+                    }
+                },
```
So between the epoch boundary and successful refresh, stale protocol params would be used by eval? And the refresh might only practically fail if a node is down?

ginnun replied:
@michalrus Yes that is correct. Alternatively we can fail (or wait) evaluate calls on epoch boundary but that might be fragile. Another (bad) option is to never cache those values.

michalrus commented on crates/node/src/chain_config_watch.rs:
```diff
@@ -0,0 +1,233 @@
+        let (tx, rx) = watch::channel(None);
+
+        tokio::spawn(async move {
+            monitor_loop(node_pool, tx).await;
```
Can this panic? 👀 It seems rather crucial for correct eval results, and we wouldn't restart it then, no?

ginnun replied:
There is nothing that can panic meaningfully at the moment (only setting slot=0 will panic). But for future code changes that's a possibility, I've added a supervisor loop in 70480e66a7218e2a63c8cda865242c72fbee1f91

michalrus commented on crates/node/src/chain_config_watch.rs:
```diff
@@ -0,0 +1,233 @@
+                    tracing::error!(
+                        "ChainConfigWatch: watch channel closed before configuration became available: {err}"
+                    );
+                    break;
```
But then we also return `()` from here, and the caller (`initialize_app` in `crates/integration_tests/src/platform/mod.rs`) would continue? This is only used in tests?

ginnun replied:
Only for testing, should not harm anything. I did add feature gating since I am here and also changed the returning strategy in cde811b9f474d05c456caa756cfb6580f43b20a8

michalrus commented on crates/integration_tests/tests/api/utils/txs/evaluate/root.rs:
```diff
@@ -0,0 +1,286 @@
+    use axum::{body::Body, http::Request};
+    use reqwest::Method;
+
+    #[tokio::test]
```
I noticed that the new integration tests are missing `#[ntest::timeout]` like other integration tests, but I don't know if they could hang? The other ones can, and that's why we added this attribute

ginnun replied:
+1, done in d2bb5c1e178d53cb81bc0ffc000fb3feda4f902b

michalrus commented on crates/integration_tests/tests/api/utils/txs/evaluate/root.rs:
```diff
@@ -0,0 +1,286 @@
+            .await
+            .expect("Failed to read response body");
+
+        let body_json: serde_json::Value = serde_json::from_slice(&body_bytes).unwrap();
```
The tests look pretty repetitive, wouldn't it be more readable to have a function that implements each one? Or even `#[rstest]` and `#[case(…)]` like e.g. here:

https://github.com/blockfrost/blockfrost-platform/blob/3bd7bfde4d1a6273c8cc58c7e43aa8c769567914/crates/platform/src/payment_cred.rs#L150-L158

ginnun replied:
I did some tidying up in d2bb5c1e178d53cb81bc0ffc000fb3feda4f902b

michalrus commented on crates/integration_tests/tests/api/utils/txs/evaluate/utxos.rs:
```diff
@@ -0,0 +1,140 @@
+    /// This test is identical to the Blockfrost test:
+    /// https://github.com/blockfrost/blockfrost-tests/blob/dc33312126e9d7c49836d7605ab72224a337bc91/src/fixtures/preview/utils/txs-evaluate-utxos.ts#L5
+    /// https://github.com/blockfrost/blockfrost-tests/blob/dc33312126e9d7c49836d7605ab72224a337bc91/src/fixtures/preview/utils/txs-evaluate-utxos.ts#L22
+    #[tokio::test]
```
Same comments – missing `#[ntest::timeout]` + repetitveness (`#[rstest]` maybe?)

ginnun replied:
+1, done in d2bb5c1e178d53cb81bc0ffc000fb3feda4f902b

michalrus commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,425 @@
+
+    /// Get or init testgen, re-spawning if config changed. Returns 503 if
+    /// chain config is not yet available
+    async fn ensure_testgen(&self) -> Result<Testgen, BlockfrostError> {
```
I'm not sure that I like that the `exec()` syscall for starting a subprocess is potentially triggered by each end user request 🤔 A small bug later on, and they can DoS our server?

Why don’t we run and control it in the background, in some Tokio thread that also monitors the output of `ChainConfigWatch`, and here just _talk_ to it? This would seem less DoS-prone

ginnun replied:
Moved the (re)spawn loop into a background testgen_supervisor task that only fires on `Arc<ChainConfigCache>` changes. Request path is now a lock-free `watch::Receiver::borrow()` See 2df81e6c

michalrus commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,425 @@
+
+        tracing::info!("ExternalEvaluator: spawning testgen-hs");
+
+        let testgen = spawn_and_init_testgen(&current_config).await.map_err(|e| {
```
I missed that in the previous review, but why is it calling `std::thread` instead of spawning a Tokio task?

ginnun replied:
Since this workload is a blocking indefinitely-long loop, I intentionally went with a dedicated thread. Of course it's up to debate, but tokio documentation suggests similar: https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html#when-to-use-spawn_blocking-vs-dedicated-threads

The supervisor task itself is `tokio::spawn` though, and the async caller boundary is `tokio::sync::mpsc`. (see 2df81e6c1148ae66569819bff171ce4e69eae3bc for latest)

michalrus commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,425 @@
+            ))
+        })?;
+
+        state.testgen = Some(testgen.clone());
```
Where is the old process being killed, after a config update from `ChainConfigWatch`?

ginnun replied:
Graceful drain. Publishing a new `Testgen` into the watch slot drops the old reference; in-flight request clones keep the worker alive until they finish, then the mpsc receiver disconnects and `testgen::spawn_child` runs `child.kill() + wait()` ([testgen.rs:294](https://github.com/blockfrost/blockfrost-platform/blob/feat/tx-evaluation-rebased/crates/testgen/src/testgen.rs#L294)).

michalrus commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,425 @@
+        additional_utxos: Option<AdditionalUtxoSet>,
+    ) -> Result<serde_json::Value, BlockfrostError> {
+        // Pre-Alonzo tx (3-element CBOR array) cannot be evaluated.
+        if tx_cbor_binary.first() == Some(&0x83) {
```
Are you 100% sure that only pre-Alonzo starts with `0x83`? And that no pre-Alonzo can start with something else?

ginnun replied:
Yes for sure. I checked the CDDLs again.
But for clearness, I changed `== Some(&0x83)` to `!= Some(&0x84)`. Also added a reminder TODO for dijkstra. It will need lots of changes including pallas anyways. You can see in 7a8f22ae655d7387d5c40fba501417c1640da770

michalrus commented on crates/tx_evaluator/src/external.rs:
```diff
@@ -0,0 +1,425 @@
+        additional_utxos: Option<Vec<AdditionalUtxoV6>>,
+    ) -> Result<serde_json::Value, BlockfrostError> {
+        // Pre-Alonzo tx (3-element CBOR array) cannot be evaluated.
+        if tx_cbor_binary.first() == Some(&0x83) {
```
Same question

ginnun replied:
Yes for sure. I checked the CDDLs again.
But for clearness, I changed `== Some(&0x83)` to `!= Some(&0x84)`. Also added a reminder TODO for dijkstra. It will need lots of changes including pallas anyways. You can see in 7a8f22ae655d7387d5c40fba501417c1640da770

michalrus commented on crates/tx_evaluator/src/helper.rs:
```diff
@@ -0,0 +1,93 @@
+/// - Non-ASCII bytes (or empty) → treat as raw binary CBOR
+/// - ASCII text → try hex decode, then base64 decode
+/// - If neither works → error
+pub fn resolve_tx_body(body: &[u8]) -> Result<Vec<u8>, String> {
```
We have this function somewhere else already, no? I remember writing it, hmm

Edit: here:

https://github.com/blockfrost/blockfrost-platform/blob/3bd7bfde4d1a6273c8cc58c7e43aa8c769567914/crates/common/src/helpers.rs#L4-L20

ginnun replied:
Close but not the same.

This one has `Base64` support and returns `Result` to surface decoding failures. I didn't reuse `binary_or_hex_heuristic` so the `submit` endpoint can keep its infallible contract while `evaluate` surfaces decode errors as 400. Not sure if submit endpoint should have base64 support. Happy to extract a shared inner hex-detection helper (and use it, tho not my liking) or use one implementation with `Base64` support, if you'd prefer.

michalrus commented on crates/tx_evaluator/src/helper.rs:
```diff
@@ -0,0 +1,93 @@
+/// - ASCII text → try hex decode, then base64 decode
+/// - If neither works → error
+pub fn resolve_tx_body(body: &[u8]) -> Result<Vec<u8>, String> {
+    let is_text = !body.is_empty() && body.iter().all(|b| b.is_ascii());
```
Is empty CBOR valid CBOR?

ginnun replied:
No, empty isn't valid CBOR. But hosted Blockfrost apis accepts an empty body and surfaces it as an Ogmios deserialisation fault. We match that by passing empty to downstream which produces the same shape of error. Updated the doc to be honest about that in fb7d63d632199358fbc24d27c5db3789ac9bd54c

michalrus commented on crates/tx_evaluator/src/native.rs:
```diff
@@ -0,0 +1,954 @@
+) -> Result<MultiEraProtocolParameters, BlockfrostError> {
+    // Create a Conway protocol parameters struct from the ProtocolParam and GenesisConfig
+    let conway_pp: ConwayProtParams = ConwayProtParams {
+        minfee_a: pp.minfee_a.required("minfee_a")? as u32,
```
Why are they `u64` in `CurrentProtocolParam`, but `u32` in `ConwayProtParams`? Are these multiple conversions always safe?

ginnun replied:
For `minfee_*` fields:
In pratice this is safe for now. But depending on CDDL(`uint`) and the ledger(`Word64`) code, it should be `u64`. I'm not sure why upstream `pallas` uses `u32`. I changed the `pallas` part in my `pallas` PR since actual data can be bigger. `CurrentProtocolParam` uses correct types based on CDDL.

For other casted fields:
These are completely safe to cast into `u32`. CDDL defines `uint4/8`  and the ledger code uses `Word16/32`. I am not sure again, why `pallas` upstream uses `u64`. Since this won't create a problem ever, I didn't modify `pallas`.

Please see c26479f542ae74bb00e675425fa6482a13673df6

michalrus commented on crates/tx_evaluator/src/ogmios5_response.rs:
```diff
@@ -0,0 +1,503 @@
+                let tx_id = utxo
+                    .get("transaction")
+                    .and_then(|transaction| transaction.get("id"))
+                    .unwrap_or(&Value::Null);
```
Is the `null` okay here, or should we error?

ginnun replied:
Never gonna happen but if happens, this is ok. I think returning the root error is more meaningful than erroring out because of this unwrap.

michalrus commented on crates/tx_evaluator/README.md:
```diff
@@ -0,0 +1,151 @@
+- Native is based on `pallas-validate`.
+- External is using `ledger` codebase via `testgen` Haskell binary.
+
+`native` produces different evaluation results when compared to `external` at the moment. Knowing that `external` depends on the ledger codebase, we can consider `external` results accurate.
```
But we don’t implement the mempool part there? 👀

ginnun replied:
Yes, no mempool handling. Haskell part doesn't fetch any data from any source. By design, it's the Rust parts that needs to fetch any data and pass it to forward.

michalrus commented on crates/tx_evaluator/src/wrapper.rs:
```diff
@@ -0,0 +1,242 @@
+    if let Some(first_result) = results.first() {
+        matches!(first_result, TxEvalResultV6::Success(_))
+    } else {
+        true
```
Can you explain this `true` here? Why are empty `results` a success?

ginnun replied:
Good catch. The empty case is produced upstream in `external.rs` when the tx has no redeemers (no Plutus scripts to evaluate); we short-circuit testgen-hs and return []. I added clarification in 7d51d8cc972a29762e00342f007b049c4691a96e

michalrus commented on crates/tx_evaluator/src/model/api.rs:
```diff
@@ -0,0 +1,1169 @@
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum EvaluationError {
+    Evaluation(serde_json::Value),
+    Deserialization(DeserializationErrorData),
+}
```
Since this is `#[serde(untagged)]`, and the first variant is arbitrary JSON, the deserializer would never reach the second variant 🤔 Do we ever use the deserializer? If not, maybe let's drop it?

ginnun replied:
Ah yes, only `Evaluation` variant is used to/from JSON. But `Deserialization` varian is also used, only to JSON. So both variant is used, but it's confusing. Hmmm.

michalrus reviewed (commented):
A couple questions – the most important one is about the mempool omission:

ginnun committed:
fix(node): restart ChainConfigWatch monitor_loop on panic

ginnun committed:
fix(tx_evaluator): update pallas; drop minfee_a/b u32 truncation; populate failure_message

ginnun committed:
docs(tx_evaluator): note why is_success_v6 returns true for empty input

ginnun committed:
test(tx_evaluator): add timeouts and consolidate evaluate integration tests

ginnun committed:
refactor(node): gate ChainConfigWatch::wait_ready behind a feature for testing-only

ginnun committed:
refactor(tx_evaluator): allow Alonzo+ CBOR prefix for evaluation

ginnun committed:
docs(tx_evaluator): clarify resolve_tx_body behavior

ginnun committed:
refactor(tx_evaluator): manage testgen-hs lifecycle via background supervisor

ginnun committed:
todo(tx_evaluator): Add issue link to spawn_and_init_testgen todo

ginnun committed:
Merge branch 'main' into feat/tx-evaluation-rebased
